xmpp-vala/src/module/xep/0047_in_band_bytestreams.vala

```diff
@@ -0,0 +1,437 @@
+                    conn.trigger_write_callback();
+                } else {
+                    conn.set_error("connection failed");
+                }
+            });
+        } else {
+            stream.get_flag(Flag.IDENTITY).add_connection(conn);
+        }
+        return conn;
+    }
+
+    void throw_if_error() throws IOError {
+        if (state == ERROR) {
+            throw new IOError.FAILED(error);
+        }
+    }
+
+    void throw_if_closed() throws IOError {
+        throw_if_error();
+        if (state == DISCONNECTING || state == DISCONNECTED) {
+            throw new IOError.CLOSED("can't read/write on a closed connection");
+        }
+    }
+
+    void set_error(string error) {
+        if (state != WAITING_FOR_CONNECT && state != DISCONNECTING && state != DISCONNECTED && state != ERROR) {
+            close_async.begin();
+        }
+        state = ERROR;
+        this.error = error;
+        stream.get_flag(Flag.IDENTITY).remove_connection(this);
+    }
+
+    public void handle_open(XmppStream stream, StanzaNode open, Iq.Stanza iq) {
+        assert(state == WAITING_FOR_CONNECT);
+        int block_size = open.get_attribute_int("block-size");
+        string? stanza = open.get_attribute("stanza");
+        if (block_size < 0 || (stanza != null && stanza != "iq" && stanza != "message")) {
+            set_error("invalid open");
+            stream.get_module(Iq.Module.IDENTITY).send_iq(stream, new Iq.Stanza.error(iq, new ErrorStanza.bad_request("missing block_size or invalid stanza")));
+            return;
+        }
+        if (stanza != null && stanza != "iq") {
+            set_error("invalid open");
+            stream.get_module(Iq.Module.IDENTITY).send_iq(stream, new Iq.Stanza.error(iq, new ErrorStanza.feature_not_implemented("cannot use message stanzas for IBB")));
+            return;
+        }
+        if (block_size > this.block_size) {
+            set_error("invalid open");
+            stream.get_module(Iq.Module.IDENTITY).send_iq(stream, new Iq.Stanza.error(iq, new ErrorStanza.build(ErrorStanza.TYPE_CANCEL, ErrorStanza.CONDITION_RESOURCE_CONSTRAINT, "opening a connection with a greater than negotiated/acceptable block size", null)));
+            return;
+        }
+        this.block_size = block_size;
+        state = CONNECTED;
+        stream.get_module(Iq.Module.IDENTITY).send_iq(stream, new Iq.Stanza.result(iq));
+        trigger_write_callback();
+    }
+    public void handle_data(XmppStream stream, StanzaNode data, Iq.Stanza iq) {
+        assert(state == CONNECTED);
+        if (input_closed) {
+            set_error("unexpected data");
+            stream.get_module(Iq.Module.IDENTITY).send_iq(stream, new Iq.Stanza.error(iq, new ErrorStanza.not_allowed("unexpected data")));
+            return;
+        }
+        int seq = data.get_attribute_int("seq");
+        // TODO(hrxi): return an error on malformed base64 (need to do this
```

Comment:
You can check that `content.length` = `base64.length`/4*3 - `number of trailing =` to verify that the base64 string was fully decoded and as such valid.

**Not** blocking for merging.

Reply:
Not yet done, seemed harder than the proposed method (whitespace is valid in base64).